### PR TITLE
[Draft] Proof of concept: Arduino framework support

### PR DESCRIPTION
**Pull request by h3ndrik**
I did a quick proof of concept to get the local wake word detection running on my ESP32 (without the S3) board. I also use the media_player coponent so I had to rewrite it to use the Arduino framework instead of the IDF.

I started removing the dependencies on the ADF, analogous to my PR in esphome: https://github.com/esphome/esphome/pull/5613

(That PR is also still a draft, since doing Voice Activity Detection is hard without proper signal processing. Just relying on the loudness doesn't work reliably and I've been waiting on something like this to get to something that is actually working.)

Unfortunately espressif's fork of tensorflow micro is only available as an IDF component. I found [an (outdated) project](https://github.com/Nickjgniklu/ESP_TF) that adapts that to an Arduino library. I made a fork and updated it: https://github.com/h3ndrik/ESP_TF

This isn't ready to merge yet. I'd like to receive some comments on the approach before cleaning it up and adding a few conditions so it either pulls in the IDF component or the Arduino library.

Btw: How do you measure performance? The debug component shows a loop time of 50-60ms on a plain ESP32.

### CODE REVIEW

**Comment by kahrendt** [comment]
Thank you for taking a look at this! I don't use the Arduino framework, but it would be nice to avoid using Espressif's ADF component.

Using the debug component, I typically get a max loop time of less than 20 ms on an ESP32-S3. A plain ESP32 may not be fast enough as the [esp-nn optimizations](https://github.com/espressif/esp-nn) show better performance increases with the S3. You can directly see the inference time by reducing the log level in the ``perform_streaming_inference_`` function. If you just increase the overall level to verbose, the extra logging may slow things down too much. When I experimented with different models, I found the maximum usable inference time is around 13 ms. After that, the ESP32 can't consistently process each new 20 ms of audio samples while doing the other necessary functions to run. This resulted in dropped audio samples, and the wake word detection was almost impossible.

**Comment by h3ndrik** [comment]
It says: Streaming Inference Latency=17 ms
The loop time varies (usually) between 32ms and 60ms, initially it is as high as 8800ms.

Wake word detection seems to work somewhat, ~~but it only gets it every other time~~ [Edit: seems pretty reliable, I initially got the mic gain too high]. I'm not sure if it drops frames, I don't think so. I'm still messing with the audio code so I just had a glance at the ML stuff. Maybe we can make the model smaller for the ESP32 without the S3, seems the tensorflow speech model has fewer parameters anyways?!

The device I'm currently testing on is an old [TAudio](https://github.com/LilyGO/TTGO-TAudio) development board. It has a plain ESP32-WROVER module with 4MB PSRAM. And it doesn't have a MEMS microphone, but a seperate codec.

I was aware of the state of audio support in esphome. But holy cow is this a mess. I want the media_player component. But the library containing the codecs and processing is Arduino only. And esphome maintains an [own fork](https://github.com/esphome/ESP32-audioI2S). Well seems they don't really maintain it, because it has not upstreamed fixes and is years out of date. And no matter if I choose the IDF or Arduino framework, I lose some of the functionality and it's just too much to do.

So, I'm not sure where I'm going with this. I still think we should get local wake word detection (with an open approach, so not the proprietary ESP-SR) into esphome. And the I2S media_player working on both supported frameworks and every microcontroller that is capable. I'm going to continue fiddling around with this but I'm not sure I can finish this properly.

**Comment by kahrendt** [comment]
That is good to know the inference time is 17 ms on a regular ESP32! I think we can experiment with the model's hyper-parameters to possibly lower it further, but that may have too large of an effect on the accuracy. In my experiments with other models, the accuracy drops quickly with less deep models or models with fewer parameters. They still work, but they have an unusable amount of false positives (who wants a voice assistant that activates 5 times when watching 1 TV episode?) An Xception neural network probably would be faster all around with only a small decrease in accuracy. After I get a better training framework setup, I will hopefully have some time to experiment with that.

The voice assistant component only buffers 32 ms of incoming audio samples (at least as currently configured), so if the max loop time is over that, I would guess some audio samples are lost. The wake word detection can still work, but I had a lot of false rejections when the max loop time is that high.

**Comment by h3ndrik** [comment]
> The voice assistant component only buffers 32 ms of incoming audio samples (at least as currently configured)

Why is that? I think the ring_buffer is 1 second long and the wake word detection seems to be running off the ring_buffer.
https://github.com/kahrendt/esphome-on-device-wake-word/blob/b67ddae5734958a5bb74b7eb7678bea3ae3ea884/esphome/components/voice_assistant/voice_assistant.cpp#L22

**Comment by sebirdman** [comment]
I did a few small tweaks on this. Really great stuff! The ESP-ADF is pretty restrictive so i've been following the arduino work and path for a little bit. Works great on my esp32-s3 with a custom microphone and speaker setup. 

I noticed that when playing music through the media_player that i've got a pretty choppy audio. I can add some checks to make sure the voice assistant stops doing detection during this, but it's not ideal. I wonder if we can filter out the speaker output somehow? 

Either way, way better than the current need to bring your board into esp-adf. Great work @h3ndrik

**Comment by kahrendt** [comment]
> > The voice assistant component only buffers 32 ms of incoming audio samples (at least as currently configured)
> 
> Why is that? I think the ring_buffer is 1 second long and the wake word detection seems to be running off the ring_buffer.
> 
> https://github.com/kahrendt/esphome-on-device-wake-word/blob/b67ddae5734958a5bb74b7eb7678bea3ae3ea884/esphome/components/voice_assistant/voice_assistant.cpp#L22

You are correct, my apologies. I do think it could still be dropping some audio samples. Try adding
``` 
ESP_LOGD(TAG, "slices available: %u", xStreamBufferBytesAvailable(this->ring_buffer_)/(2*NEW_SAMPLES_TO_GET));
```
to the ``slice_available_()`` function. If the component is keeping up, it should output 0 or 1. If it is consistently higher than that, then it is most likely going to fill up the entire buffer. With the buffer full, some audio samples will be dropped reducing accuracy (and this will also greatly increase the latency for detecting the word).

I merged jesserockz's PR this morning, and the code now uses the FreeRTOS ring buffer like you do here. So to add Arduino support, it seems we may only need to add the Arduino library code to ``__init__.py``?

**Comment by h3ndrik** [comment]
Oh wow, things are happening. I was away from my keyboard for a few days. Guess I have to rework a few of my own requests. Sure, most of my changes seem superseded by [jesserockz](https://github.com/jesserockz)' https://github.com/esphome/esphome/pull/6102 Seems it's now just adding the library. I'll test this and also the performance of my ESP32.

Last thing is to decide what to do with the library. I just forked it and made some changes because it was outdated. Maybe the original author (it's just a wrapper) wants to maintain it or esphome wants to maintain an own fork. We'd have to talk to a few people to come up with a sustainable solution. Maybe I should send a pull request there and we can also talk to the esphome people if you submit this to get incorporated into esphome.

**Comment by alior101** [comment]
Hi @h3ndrik !
What's the status of this PR ? I'm interested in getting rid of the ADF also since I'm using an S3 board without any codec and the main branch crashes on audio board init...

[15:00:15][I][esp_adf:019]: Start codec chip
[15:00:15][D][esp-idf:000]: E (218) AUDIO_BOARD: board.c:45 (audio_board_init): Memory exhausted

**Comment by kahrendt** [comment]
Thanks for getting it to work with Arduino and for testing things out! As this is now officially part of ESPHome, Arduino support should be considered as a PR to the ESPHome project.
